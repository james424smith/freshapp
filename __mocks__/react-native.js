import * as ReactNative from "react-native";

export * from "react-native";
export const Vibration = { vibrate: jest.fn() };
export const Adnimated = { spring: jest.fn() };

export default Object.setPrototypeOf(
  {
    Vibration,
    Adnimated,
  },
  ReactNative
);
