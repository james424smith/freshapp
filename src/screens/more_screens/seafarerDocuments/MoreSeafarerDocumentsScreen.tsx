import React, { useState, useEffect } from "react";
import { ScrollView, StyleSheet, View } from "react-native";
import { SeafarerDocumentsFile } from "../../../interfaces";
import { useSelector, useDispatch } from "react-redux";
import { getSeafarerDocumentsFile } from "../../../redux/actions";
import ImageRender from "../../../components/ImgRender";
import * as RNFS from "react-native-fs";
import LoadingScreen from "../../../components/LoadingScreen";

import { IRootReducerType } from "../../../redux/reducers";
import PdfRender from "../../../components/PdfRender";
import { DocumentsStackParamList } from "../../../navigation/MoreScreenNavigation";
import { RouteProp } from "@react-navigation/native";

interface Props {
  route: RouteProp<DocumentsStackParamList, "SeafarerDocumentsFile">;
}

const styles = StyleSheet.create({
  scrollViewContainer: {
    flexGrow: 1,
  },
  scrollView: {
    flex: 1,
  },
  root: { flex: 1 },
  container: { alignSelf: "center", marginTop: 50 },
  centerIndicator: {
    flex: 1,
    marginBottom: 50,
    alignItems: "center",
    justifyContent: "center",
  },
});
const logoSize = 100;

const SeafarerDocumentsFileScreen = (props: Props) => {
  const [documentFromLocal, setDocumentFromLocal] = useState<
    SeafarerDocumentsFile | undefined
  >(undefined);
  const { route } = props;

  const seafarerDocumentsFile = useSelector<
    IRootReducerType,
    SeafarerDocumentsFile | undefined
  >(({ seafarerDocuments }) => seafarerDocuments.seafarerDocumentsFile);
  const isLoadingNew = useSelector<IRootReducerType, boolean>(
    ({ seafarerDocuments }) => seafarerDocuments.isLoadingNew
  );
  const dispatch = useDispatch();
  const openFileFromLocalStorage = (
    documentId: string | number,
    documentCounter: number,
    fileType: string
  ) => {
    return `${RNFS.DocumentDirectoryPath}/document/id-${documentId}-counter-${documentCounter}.${fileType}`;
  };

  useEffect(() => {
    const documentDetails = route.params?.documentDetails;
    //if document switched to offline
    if (documentDetails.isOffline) {
      const { documentId, documentCounter, fileType } = documentDetails;
      //open the offline document from the path
      RNFS.readFile(
        openFileFromLocalStorage(documentId, documentCounter, fileType)
      )
        .then((file: any) => {
          setDocumentFromLocal({ document: file, fileType });
        })
        .catch((error: any) => {
          console.log(error);
        });
      //If not offline, fetch the document from API
    } else {
      dispatch(getSeafarerDocumentsFile(documentDetails, false));
    }
  }, [dispatch, route.params]);

  const chooseComponentToShow = (file: SeafarerDocumentsFile) => {
    return file.fileType === "PDF" ? (
      <PdfRender document={file.document} resourceType={"base64"} />
    ) : (
      <ImageRender
        document={file.document}
        resourceType={"base64"}
        fileType={file.fileType}
      />
    );
  };

  const itemToShow = (
    isLoadingNewLocal: boolean,
    imgFile?: SeafarerDocumentsFile
  ) => {
    if (isLoadingNewLocal) {
      return (
        <View style={styles.root}>
          <LoadingScreen logoWidthSize={logoSize} logoHeightSize={logoSize} />
        </View>
      );
    } else {
      return imgFile ? chooseComponentToShow(imgFile) : undefined;
    }
  };

  //checking if the document is offline.
  const img = documentFromLocal ? documentFromLocal : seafarerDocumentsFile;
  return (
    <ScrollView
      style={styles.scrollView}
      contentContainerStyle={styles.scrollViewContainer}
    >
      {itemToShow(isLoadingNew, img)}
    </ScrollView>
  );
};

export default SeafarerDocumentsFileScreen;
