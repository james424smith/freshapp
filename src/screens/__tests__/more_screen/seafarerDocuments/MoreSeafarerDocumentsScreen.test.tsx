import React from "react";
import MoreSeafarerDocumentsScreen from "../../../more_screens/seafarerDocuments/MoreSeafarerDocumentsScreen";
import { render, cleanup, waitFor } from "@testing-library/react-native";
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";
import RNFS from "react-native-fs";
const document = {
  documentId: 1234,
  documentCounter: 12,
  docLicNumber: "1234",
  nation: "CY",
  description: "description",
  expiryDate: "01-01-2021",
  issueDate: "01-01-2019",
  docFileSize: "1mb",
  isDeactivated: false,
  categoryNumber: "4",
  fileType: "PDF",
  isExpired: false,
  isOffline: false,
  documentExtension: "pdf",
  marked: false,
  hasAttachment: true,
};

describe("MoreSeafarerDocumentsScreen snapshot test", () => {
  afterEach(() => {
    cleanup();
    jest.clearAllMocks();
    jest.restoreAllMocks();
  });

  it("should render correct ui for MoreSeafarerDocumentsScreen when payslip is present and downloaded and loading is false", async () => {
    jest.useRealTimers();
    const route = {
      key: "SeafarerDocumentsFile-wPFrEJG5kjiOrE_Qk2kVs",
      name: "SeafarerDocumentsFile" as "SeafarerDocumentsFile",
      params: {
        documentDetails: { ...document, isOffline: true },
      },
    };
    const mockedState = {
      seafarerDocuments: {
        seafarerDocumentsFile: {
          document: "1234",
          fileType: "pdf",
        },
        isLoadingNew: false,
      },
      network: { isConnected: true },
    };
    const mockedStore = configureMockStore();

    const { toJSON } = render(
      <Provider store={mockedStore(mockedState)}>
        <MoreSeafarerDocumentsScreen route={route} />
      </Provider>
    );
    await waitFor(() => {
      expect(toJSON()).toMatchSnapshot();
    });
  });

  it("should render correct ui for MoreSeafarerDocumentsScreen when payslip is present and downloaded and loading is false and filetype is image", async () => {
    jest.useRealTimers();
    const route = {
      key: "SeafarerDocumentsFile-wPFrEJG5kjiOrE_Qk2kVs",
      name: "SeafarerDocumentsFile" as "SeafarerDocumentsFile",
      params: {
        documentDetails: { ...document, isOffline: true, fileType: "image" },
      },
    };
    const mockedState = {
      seafarerDocuments: {
        seafarerDocumentsFile: {
          document: "1234",
          fileType: "pdf",
        },
        isLoadingNew: false,
      },
      network: { isConnected: true },
    };
    const mockedStore = configureMockStore();

    const { toJSON } = render(
      <Provider store={mockedStore(mockedState)}>
        <MoreSeafarerDocumentsScreen route={route} />
      </Provider>
    );
    await waitFor(() => {
      expect(toJSON()).toMatchSnapshot();
    });
  });

  it("should render correct ui for MoreSeafarerDocumentsScreen when payslip is present and downloaded and loading is false and readFile fails", async () => {
    jest.useRealTimers();
    const spy = jest.spyOn(RNFS, "readFile").mockRejectedValue({});
    const route = {
      key: "SeafarerDocumentsFile-wPFrEJG5kjiOrE_Qk2kVs",
      name: "SeafarerDocumentsFile" as "SeafarerDocumentsFile",
      params: {
        documentDetails: { ...document, isOffline: true },
      },
    };
    const mockedState = {
      seafarerDocuments: {
        seafarerDocumentsFile: {
          document: "1234",
          fileType: "pdf",
        },
        isLoadingNew: false,
      },
      network: { isConnected: true },
    };
    const mockedStore = configureMockStore();

    const { toJSON } = render(
      <Provider store={mockedStore(mockedState)}>
        <MoreSeafarerDocumentsScreen route={route} />
      </Provider>
    );
    await waitFor(() => {
      expect(toJSON()).toMatchSnapshot();
    });
    spy.mockRestore();
  });

  it("should render correct ui for MoreSeafarerDocumentsScreen when payslip is present and not downloaded and loading is false", () => {
    const route = {
      key: "SeafarerDocumentsFile-wPFrEJG5kjiOrE_Qk2kVs",
      name: "SeafarerDocumentsFile" as "SeafarerDocumentsFile",
      params: {
        documentDetails: document,
      },
    };
    const mockedState = {
      seafarerDocuments: {
        seafarerDocumentsFile: {
          document: "1234",
          fileType: "pdf",
        },
        isLoadingNew: false,
      },
      network: { isConnected: true },
    };
    const mockedStore = configureMockStore();

    const { toJSON } = render(
      <Provider store={mockedStore(mockedState)}>
        <MoreSeafarerDocumentsScreen route={route} />
      </Provider>
    );

    expect(toJSON()).toMatchSnapshot();
  });

  it("should render correct ui for MoreSeafarerDocumentsScreen when payslip is undefined and loading is false", () => {
    const route = {
      key: "SeafarerDocumentsFile-wPFrEJG5kjiOrE_Qk2kVs",
      name: "SeafarerDocumentsFile" as "SeafarerDocumentsFile",
      params: {
        documentDetails: document,
      },
    };
    const mockedState = {
      seafarerDocuments: {
        seafarerDocumentsFile: undefined,
        isLoadingNew: false,
      },
      network: { isConnected: true },
    };
    const mockedStore = configureMockStore();

    const { toJSON } = render(
      <Provider store={mockedStore(mockedState)}>
        <MoreSeafarerDocumentsScreen route={route} />
      </Provider>
    );

    expect(toJSON()).toMatchSnapshot();
  });
  it("should render correct ui for MoreSeafarerDocumentsScreen when payslip is present and loading is true", () => {
    const route = {
      key: "SeafarerDocumentsFile-wPFrEJG5kjiOrE_Qk2kVs",
      name: "SeafarerDocumentsFile" as "SeafarerDocumentsFile",
      params: {
        documentDetails: document,
      },
    };
    const mockedState = {
      seafarerDocuments: {
        seafarerDocumentsFile: { document: "1234", fileType: "pdf" },
        isLoadingNew: true,
      },
      network: { isConnected: true },
    };
    const mockedStore = configureMockStore();

    const { toJSON } = render(
      <Provider store={mockedStore(mockedState)}>
        <MoreSeafarerDocumentsScreen route={route} />
      </Provider>
    );

    expect(toJSON()).toMatchSnapshot();
  });
});
