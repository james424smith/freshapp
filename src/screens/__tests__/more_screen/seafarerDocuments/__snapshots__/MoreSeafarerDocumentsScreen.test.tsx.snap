// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoreSeafarerDocumentsScreen snapshot test should render correct ui for MoreSeafarerDocumentsScreen when payslip is present and downloaded and loading is false 1`] = `
<RCTScrollView
  contentContainerStyle={
    Object {
      "flexGrow": 1,
    }
  }
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View>
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <PDFView
        fadeInDuration={10}
        onError={[Function]}
        onLoad={[Function]}
        onPageChanged={[Function]}
        onScrolled={[Function]}
        resource=""
        resourceType="base64"
        style={
          Object {
            "flex": 1,
            "width": 750,
          }
        }
        textEncoding="utf-8"
        urlProps={Object {}}
      />
    </View>
  </View>
</RCTScrollView>
`;

exports[`MoreSeafarerDocumentsScreen snapshot test should render correct ui for MoreSeafarerDocumentsScreen when payslip is present and downloaded and loading is false and filetype is image 1`] = `
<RCTScrollView
  contentContainerStyle={
    Object {
      "flexGrow": 1,
    }
  }
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View>
    <RCTScrollView
      contentContainerStyle={
        Object {
          "flexGrow": 1,
        }
      }
      maximumZoomScale={2.5}
      minimumZoomScale={1}
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View>
        <Image
          resizeMode="contain"
          source={
            Object {
              "uri": "data:application/image;base64,",
            }
          }
          style={
            Object {
              "flex": 1,
              "width": 750,
            }
          }
        />
      </View>
    </RCTScrollView>
  </View>
</RCTScrollView>
`;

exports[`MoreSeafarerDocumentsScreen snapshot test should render correct ui for MoreSeafarerDocumentsScreen when payslip is present and downloaded and loading is false and readFile fails 1`] = `
<RCTScrollView
  contentContainerStyle={
    Object {
      "flexGrow": 1,
    }
  }
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View>
    <RCTScrollView
      contentContainerStyle={
        Object {
          "flexGrow": 1,
        }
      }
      maximumZoomScale={2.5}
      minimumZoomScale={1}
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View>
        <Image
          resizeMode="contain"
          source={
            Object {
              "uri": "data:application/pdf;base64,1234",
            }
          }
          style={
            Object {
              "flex": 1,
              "width": 750,
            }
          }
        />
      </View>
    </RCTScrollView>
  </View>
</RCTScrollView>
`;

exports[`MoreSeafarerDocumentsScreen snapshot test should render correct ui for MoreSeafarerDocumentsScreen when payslip is present and loading is true 1`] = `
<RCTScrollView
  contentContainerStyle={
    Object {
      "flexGrow": 1,
    }
  }
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View>
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
            "marginBottom": 112.5,
          }
        }
      >
        <Image
          source={Object {}}
          style={
            Object {
              "height": 100,
              "width": 100,
            }
          }
        />
      </View>
    </View>
  </View>
</RCTScrollView>
`;

exports[`MoreSeafarerDocumentsScreen snapshot test should render correct ui for MoreSeafarerDocumentsScreen when payslip is present and not downloaded and loading is false 1`] = `
<RCTScrollView
  contentContainerStyle={
    Object {
      "flexGrow": 1,
    }
  }
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View>
    <RCTScrollView
      contentContainerStyle={
        Object {
          "flexGrow": 1,
        }
      }
      maximumZoomScale={2.5}
      minimumZoomScale={1}
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View>
        <Image
          resizeMode="contain"
          source={
            Object {
              "uri": "data:application/pdf;base64,1234",
            }
          }
          style={
            Object {
              "flex": 1,
              "width": 750,
            }
          }
        />
      </View>
    </RCTScrollView>
  </View>
</RCTScrollView>
`;

exports[`MoreSeafarerDocumentsScreen snapshot test should render correct ui for MoreSeafarerDocumentsScreen when payslip is undefined and loading is false 1`] = `
<RCTScrollView
  contentContainerStyle={
    Object {
      "flexGrow": 1,
    }
  }
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View />
</RCTScrollView>
`;
